ouroboros-consensus-diffusion/src/ouroboros-consensus-diffusion/Ouroboros/Consensus/NodeKernel.hs

```diff
@@ -155,19 +188,59 @@ initNodeKernel
 initNodeKernel args@NodeKernelArgs { registry, cfg, tracers
                                    , chainDB, initChainDB
                                    , blockFetchConfiguration
+                                   , durationUntilTooOld
+                                   , gsmMarkerFileView
                                    } = do
     -- using a lazy 'TVar', 'BlockForging' does not have a 'NoThunks' instance.
     blockForgingVar :: LazySTM.TMVar m [BlockForging m blk] <- LazySTM.newTMVarIO []
     initChainDB (configStorage cfg) (InitChainDB.fromFull chainDB)
 
     st <- initInternalState args
+    let IS
+          { blockFetchInterface
+          , fetchClientRegistry
+          , mempool
+          , peerSharingRegistry
+          , varCandidates
+          , varIdlers
+          , varLedgerJudgement
+          } = st
+
+    do  let gsm = GSM.realGsmEntryPoints GSM.GsmView
+              { GSM.candidateOverSelection    = \(headers, _lst) candidate ->
+                    case AF.intersectionPoint headers candidate of
+                        Nothing -> GSM.CandidateDoesNotIntersect
+                        Just{}  ->
+                            GSM.WhetherCandidateIsBetter
+                          $ -- precondition requires intersection
+                            preferAnchoredCandidate
+                                (configBlock cfg)
+                                headers
+                                candidate
+              , GSM.durationUntilTooOld       = case durationUntilTooOld of
+                  Nothing -> Nothing
+                  Just wd -> Just $ \(_headers, lst) ->
+                    computeDurationUntilTooOld cfg wd lst
+              , GSM.equivalent                = (==) `on` (AF.headPoint . fst)
+              , GSM.getChainSyncCandidates    = readTVar varCandidates
+              , GSM.getChainSyncIdlers        = readTVar varIdlers
+              , GSM.getCurrentSelection       = do
+                  headers        <- ChainDB.getCurrentChain  chainDB
+                  extLedgerState <- ChainDB.getCurrentLedger chainDB
+                  return (headers, ledgerState extLedgerState)
+              , GSM.setCaughtUpPersistentMark = \upd ->
+                  (if upd then GSM.touchMarkerFile else GSM.removeMarkerFile)
```

Comment:
@bolt12 is seeing an error in his initial integration.

```
{
"app": [],
"at": "2024-01-11T07:27:57.71Z",
"data": {
"error": "ExceptionInLinkedThread \"ThreadId 46\" (FsError {fsErrorType = FsResourceDoesNotExist, fsErrorPath = /var/lib/cardano-node/db-testnet/gsm/CaughtUpMarker, fsErrorString = \"No such file or directory\", fsErrorNo = Just (Errno 2), fsErrorStack = CallStack(from HasCallStack):\n  prettyCallStack, called at src/System/FS/API/Types.hs:294:23 in fs-api-0.2.0.1-7CfJAP5WJBILIrvJjDwNip:System.FS.API.Types\n  ioToFsError, called at src/System/FS/IO.hs:90:41 in fs-api-0.2.0.1-7CfJAP5WJBILIrvJjDwNip:System.FS.IO\n  handleError,called at src/System/FS/IO.hs:86:23 in fs-api-0.2.0.1-7CfJAP5WJBILIrvJjDwNip:System.FS.IO\n  rethrowFsError, called at src/System/FS/IO.hs:68:36 in fs-api-0.2.0.1-7CfJAP5WJBILIrvJjDwNip:System.FS.IO\n  removeFile, called at src/System/FS/API.hs:130:5 in fs-api-0.2.0.1-7CfJAP5WJBILIrvJjDwNip:System.FS.API, fsLimitation = False})",
"kind": "ConsensusStartupException"
},
"env": "8.7.3:00000",
"host": "server-e",
"loc": null,
"msg": "",
"ns": [
"cardano.node"
],
"pid": "1925269",
"sev": "Critical",
"thread": "7"
}```

I'm this `removeMarkerFile` is involved somehow. (Or maybe a different one.)

Reply:
Ah, I think this is the explanation: https://github.com/IntersectMBO/ouroboros-consensus/pull/808/files#r1448966065